### PR TITLE
feat(aio): add the ability to update a PR's preview visibility

### LINKED ISSUE #16526: aio: have a way to see previews for any PR
Currently, previews are only generated for PRs that (a) touch some file inside `aio/` and (b) have a trusted author. Often it is useful to be able to see a preview for PRs that don't meet these requirements (e.g. touch API docs outside `aio/` or the changes have been verified by a trusted user).

### PR DESCRIPTION
**Pull request by gkalpak**
This PR implements the ability to update a PR's preview visibility and exposes it as an API (`/check-pr-visibility`), so that it can be used with an automatic trigger (e.g. a GitHub webhook) to instantly update a PR's preview visibility when it changes.

This builds on top of #17756 (to avoid merge conflicts).

Fixes #16526.

### CODE REVIEW

**Comment by gkalpak** [comment]
I split it up into several commits for easier review. (I can squash some of them before merging if necesary.)

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/constants.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// Constants
+/* tslint:disable: variable-name */
+export const BV_verify_error = 'FAKE_VERIFICATION_ERROR';
+export const BV_verify_verifiedNotTrusted = 'FAKE_VERIFIED_NOT_TRUSTED';
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/constants.ts]: I don't know what the official reasoning is, but I have to say that using all capitals for the names of such constants reads much easier for me.
Also, I am not sure what the `BV_` prefix is for or why it is needed.

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 // Imports
 import * as path from 'path';
+import * as c from './constants';
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts]: Why import this `as c` rather than just destructuring the constants that you want?
Is it that you envisage there being a large number of constants to import?

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts [hunk 3]
```diff
@@ -14,9 +15,6 @@ h.runForAllSupportedSchemes((scheme, port) => describe(`integration (on ${scheme
   const getFile = (pr: string, sha: string, file: string) =>
     h.runCmd(`curl -iL ${scheme}://pr${pr}-${h.getShordSha(sha)}.${host}/${file}`);
   const uploadBuild = (pr: string, sha: string, archive: string, authHeader = 'Token FOO') => {
-    // Using `FAKE_VERIFICATION_ERROR` or `FAKE_VERIFIED_NOT_TRUSTED` as `authHeader`,
-    // we can fake the response of the overwritten `BuildVerifier.verify()` method.
-    // (See 'lib/upload-server/index-test.ts'.)
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts]: Why remove this comment? If it is still accurate, then it provides some valuable reasoning for what these constants are for.
If you don't like it here then perhaps move the comment to the `constants.ts` file?

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts [hunk 4]
```diff
@@ -96,8 +114,8 @@ class UploadServerFactory {
       } else if (!archive) {
         this.throwRequestError(400, `Missing or empty '${X_FILE_HEADER}' header`, req);
       } else {
-        buildVerifier.
-          verify(+pr, authHeader).
+        Promise.resolve().
```
Comment (petebacondarwin) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts]: Why this change?

**Review by petebacondarwin — APPROVED** [review]
A couple of small nits but LGTM

**Comment by petebacondarwin** [comment]
Needs a rebase and then we can merge (unless you want to fix the nits first).

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/constants.ts (hunk 1 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/constants.ts]: WRT to casing: I was using a `<class>_<method>_<return-value>` pattern (sort of) (e.g. BV_getPrIsTrusted_` corresponds to the `BuildVerifier.prototype.getPrIsTrusted` method.

This is a whole way of faking some values in e2e-ish tests is quite sketchy anyway. Happy to discuss alternative ideas :confused:

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts (hunk 2 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts]: There are four constants now and it is not unlikely to have more in the future. They have quite long names, that's why I went with `as c`, but I don't mind changing that.

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts (hunk 3 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts]: I removed it from here, because it is not the only place where we use these constants for fake responses. I have placed comments in [other places][1], but I agree that having a comment in `constants.ts` too would be helpful.

[1]: https://github.com/angular/angular/pull/17807/files#diff-89b8f78194f3810bb2c69f2ddc345bd4R13

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts (hunk 4 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts]: Mainly uniformity with other places :grin:
It doesn't make any difference here, since `BuildVerifier#verify()` returns a promise anyway, but it sometimes helps when mocking the `verify()` method in tests, because you don't always have to return a Promise from the mock.

File: aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf [hunk 5]
```diff
@@ -88,6 +88,21 @@ server {
     resolver 127.0.0.1;
   }
 
+  # Re-check PR visibility
+  location "~^/check-pr-visibility/?$" {
```
Comment (IgorMinar) [inline aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf]: a more generic name would likely be more suitable. the webhook doesn't inform us to check the visibility, the webhook just calls us when a PR is updated, so it would be better to call the endpoint what it is - `/pr-updated` or `pr-update-notification`.

Comment (IgorMinar) [inline aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf]: in the future we could then use this endpoint for notifications of closed PRs that could be cleaned up. (no need to implement this now)

**Review by IgorMinar — COMMENTED** [review]
What do you think about the end-point rename? once we agree on this I can create the webhook for you.

**Comment by IgorMinar** [comment]
oh.. and the PR also needs a rebase because it contains already merged commits.

File: aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf (hunk 5 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf]: I was thinking this endpoint could be used for actions that might affect the PR's visibility. E.g. in the future we could have a webhook for when the members of a GitHub team are changed (i.e. adding or removing an author to/from `aio-contributors` would change the visibility of their PRs).

I guess you don't really know what an endpoint will be used for, until you use it. Happy to change that since your scenario sounds more likely to happen than mine :grin:

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/constants.ts (hunk 1 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/constants.ts]: I kept the names, but added some comments to describe what the variables are for.

File: aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts (hunk 3 above)
Comment (gkalpak) [inline aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/server-integration.e2e.ts]: Added a comment in `constants.ts`.

**Comment by gkalpak** [comment]
Rebased, adressed feedback and renamed API endpoint to `/pr-updated` (diffs: 15098a462, 403faf796, ee6365387).

**Comment by petebacondarwin** [comment]
Cherry picked to 4.2.x via

406dd45df19a1c130c36230fdbd53bf0a7b39d19 feat(aio): add API endpoint for notifying about PR updates
9948ccf36543372d0e8419c4e1d08f8837c26ef0 feat(aio): implement a way to check and update a PR\'s preview visibility
9e8d773e4e877e49e8b447b8267e15445c4c3abb refactor(aio): unify error messages for invalid requests to upload-server
aea11bca8770cae883b95bc4871e170262d105f8 refactor(aio): use dedicated `constants.ts` file for e2e-specific constants
08b4e1edc98fee276748e42bac8d3c2c35c68a6d refactor(aio): move script to a more relevant directory
